### PR TITLE
sql: dedup inverted index keys

### PR DESCRIPTION
**Pull request by m-schneider**
Before if we added a JSON columns with a duplicate key, for example
["a", "a"] the insertion would fail because index key generated
would be the same for both values resulting in a duplicate key error.

Now the whole set of keys generated for a JSON object will be deduped
so we will be able to insert and lookup JSON object with duplicate
paths.

Release note: None

### CODE REVIEW

**Comment by justinj** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3U9yWuUFjiCK0Ry2Sg:-L3U9yWuUFjiCK0Ry2Sh:bwm8ccn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/logictest/testdata/logic_test/inverted_index#L270)):*
> ```
> SELECT * from d where b @> '[]' ORDER BY a;
> 
> statement ok
> ```

Would like to see some more involved tests, in particular nested arrays.

I would be really happy to see the random JSON generator be leveraged here, testing the invariant that a JSON object should always be successfully inserted(?) we might need to augment it a bit to possibly generate duplicate strings though (doesn't have to hold up this PR).

---

*[pkg/sql/sqlbase/table.go, line 1587 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3U9XsGVt8sgvwi9vUK:-L3U9XsHiDllSjW9CGKb:b-6kdtqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/sqlbase/table.go#L1587)):*
> ```Go
> 	switch t := tree.UnwrapDatum(nil, val).(type) {
> 	case *tree.DJSON:
> 		keys, err := (t.JSON).EncodeInvertedIndexKeys(inKey)
> ```

I feel like perhaps the deduping should happen one level lower, inside of EncodeInvertedIndexKeys, since as it is now it feels like the general flow here is just that that function gives back wrong results that this function has to clean up. Don't feel super strongly though.

---

*[pkg/sql/sqlbase/table.go, line 1593 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3UAoFNVQ8MuslHkPeL:-L3UAoFNVQ8MuslHkPeM:b-ncr47z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/sqlbase/table.go#L1593)):*
> ```Go
> 		keySet := make(map[string]struct{})
> 		for _, key := range keys {
> 			keySet[string(key)] = struct{}{}
> ```

I think this setup will cause some unnecessary extra allocations, since we're going to convert from a bytestring, to a string, back to a bytestring later. Is there some way we can avoid that? I don't know how big of a deal those allocations will be but it seems like there might be some kind of easy fix (but maybe not).

One strawman would be to have these map to the original byte slices instead of `struct{}` and then in the loop below grab those instead of casting the strings back to byte slices. That would at least cut down the extra allocations from 2 to 1 (unless this is already optimized somehow by the compiler), but there might be an easy enough way to get rid of that last one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665)*

**Comment by m-schneider** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3U9yWuUFjiCK0Ry2Sg:-L3UEjydHOK8yJX823sf:b-f7s5r8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/logictest/testdata/logic_test/inverted_index#L270)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Would like to see some more involved tests, in particular nested arrays.

I would be really happy to see the random JSON generator be leveraged here, testing the invariant that a JSON object should always be successfully inserted(?) we might need to augment it a bit to possibly generate duplicate strings though (doesn't have to hold up this PR).
</blockquote></details>

I will be adding random JSON tests in a followup PR. 

---

*[pkg/sql/sqlbase/table.go, line 1587 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3U9XsGVt8sgvwi9vUK:-L3UEKNKBFlkVBiQdYvc:b-17t2u6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/sqlbase/table.go#L1587)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I feel like perhaps the deduping should happen one level lower, inside of EncodeInvertedIndexKeys, since as it is now it feels like the general flow here is just that that function gives back wrong results that this function has to clean up. Don't feel super strongly though.
</blockquote></details>

This is the only function that accesses the EncodeInvertedIndexKeys in the JSON package which is recursive. We have all of the full values here that's why I chose to do the deduping at this point. 

---

*[pkg/sql/sqlbase/table.go, line 1593 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3UAoFNVQ8MuslHkPeL:-L3UEZuthhTg6r9lpkmE:bobk8y6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/sqlbase/table.go#L1593)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think this setup will cause some unnecessary extra allocations, since we're going to convert from a bytestring, to a string, back to a bytestring later. Is there some way we can avoid that? I don't know how big of a deal those allocations will be but it seems like there might be some kind of easy fix (but maybe not).

One strawman would be to have these map to the original byte slices instead of `struct{}` and then in the loop below grab those instead of casting the strings back to byte slices. That would at least cut down the extra allocations from 2 to 1 (unless this is already optimized somehow by the compiler), but there might be an easy enough way to get rid of that last one.
</blockquote></details>

That's a very good point. We can use the unsafe library here, because we're not going to be modifying those slices.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665)*

**Comment by justinj** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3U9yWuUFjiCK0Ry2Sg:-L3UFwSIMGI6nvc-1aom:b-8sbvx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/logictest/testdata/logic_test/inverted_index#L270)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I will be adding random JSON tests in a followup PR. 
</blockquote></details>

Could you include some nested arrays in the current iteration?

---

*[pkg/sql/sqlbase/table.go, line 1593 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3UAoFNVQ8MuslHkPeL:-L3UFJaqGkSy4Bgq8WUl:bajjfhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/sqlbase/table.go#L1593)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

That's a very good point. We can use the unsafe library here, because we're not going to be modifying those slices.
</blockquote></details>

Whoa legit, I didn't know this worked :thumbsup: can you include a comment explaining the use of `unsafe` here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3UGKYN5yy_Wmajprs6:bnfp4nl)*

**Review by jordanlewis — APPROVED** [review]
LGTM. Is there a way to make this more efficient given whatever it is that we know about the sortedness of JSON datatypes?

**Comment by jordanlewis** [comment]
To be more specific, aren't object types already sorted by key in memory? Putting arrays aside for a second, if objects were all sorted by key in memory then de-duplication should be as simple as iterating through the output list discarding duplicates with a size 1 lookback.

Arrays make things more complicated, since they're not sorted. But perhaps we could sort the keys when we encounter an array. This all might be more efficient than making a giant in-memory map for all of the keys.

**Comment by justinj** [comment]
I haven't thought about it in too much detail yet, but I think the approach you're describing might be complicated by cases like this:
```
[{a:1,b:1},{b:1,c:1},{a:1,c:1}]
```

**Comment by m-schneider** [comment]
We could also sort the whole thing assuming it's already close to sorted, it shouldn't take that long and then do the size one lookup.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3UOWkQY9p4xbKJ-Y53:b-ru3ju7)*

**Comment by jordanlewis** [comment]
Wait a minute, we shouldn't be doing this globally, right? There's no sense in doing the deduplication after everything's already created. Imagine a huge JSON object with only 1 array in it, and the array is at the top level.

You should move the deduplication logic to `jsonArray.EncodeInvertedIndexKeys`. That way we never have to run deduplication unless we actually have an array that we're processing with more than 1 element in it.

**Comment by justinj** [comment]
oh yes you're absolutely right, it should be done there.

**Comment by m-schneider** [comment]
Hmm, what if you have a lot of nested arrays? It'll be reprocessing the same data multiple time.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3US-uYcKohqPZI6qHD:byqhwpr)*

**Comment by jordanlewis** [comment]
It won't be fully reprocessing the same data, though, since we know the sub-arrays will be sorted by the recursive call, right? Similar to how merge sort works.

**Comment by jordanlewis** [comment]
If you don't want to do this now, I think it's fine to merge as-is as long as you make an issue to track it.

**Comment by m-schneider** [comment]
My only concern with doing it in the array function is something like [[[[[[[[[["a", "a", "a"]]]]]]]]]]]]. Even though this is a totally contrived example, it would have to go over every "a" however deeply it's nested. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3UiGDSMuZ_mu_wWYdX:b-psq5e7)*

**Comment by jordanlewis** [comment]
Well in that case, since each outer array only has 1 element, we could skip the deduplication entirely, right?

**Comment by m-schneider** [comment]
In that case yes, but if it was [[[[[[[[[["a", "b", "c"]]]]]]]]]]]] then we couldn't.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3UjYKl7wb-GAV88DZn:bco7gb5)*

**Comment by m-schneider** [comment]
*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3U9yWuUFjiCK0Ry2Sg:-L3Ujfr3FUA4pabjj4wt:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/logictest/testdata/logic_test/inverted_index#L270)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Could you include some nested arrays in the current iteration?
</blockquote></details>

Added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665)*

**Comment by jordanlewis** [comment]
> In that case yes, but if it was [[[[[[[[[["a", "b", "c"]]]]]]]]]]]] then we couldn't.

At each outer level of recursion, the array that we're examining has 1 element. In that case, we don't have to do any work, since we know that the element has already been operated upon and therefore sorted and de-duplicated. Therefore, we may immediately prepend our array marker to each key from the pre-sorted set and output without doing any extra work.

**Comment by m-schneider** [comment]
Of course! I changed it to a merge.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3UrpgDjxum_SZxKuUP:biv11ra)*

**Comment by m-schneider** [comment]
*[pkg/sql/sqlbase/table.go, line 1593 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3UAoFNVQ8MuslHkPeL:-L3UrzPaffRtQ1cTr0Gw:b9nolnn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b5e88c9bf2506cc94957ee6b33d0d7c20dd3c5b/pkg/sql/sqlbase/table.go#L1593)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Whoa legit, I didn't know this worked :thumbsup: can you include a comment explaining the use of `unsafe` here?
</blockquote></details>

Got ride of the hashset.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665)*

File: pkg/util/json/json.go
Comment (jordanlewis) [inline pkg/util/json/json.go]: Doesn't this break the invariant that each key slice is sorted, and also leave us with duplicate keys? I could be misunderstanding, but it looks like the caller of this function creates a new key for every value in the returned slice. In that case, we'll be incorrectly emitting `nil`-suffixed keys.

**Review by jordanlewis — COMMENTED** [review]
Thanks for taking my suggestion!

I'd like to see some more test cases for this in `TestEncodeDecodeJSONInvertedIndex`. The logic tests are insufficient for testing changes to the encoding, because they don't allow us to exhaustively test that the index keys were created correctly. We're missing test cases in that test for duplicate paths.

**Comment by m-schneider** [comment]
*[pkg/util/json/json.go, line 687 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3UjSrMgJg6FhXJtlam-r3-687:-L3V0Wd17hlhrxEih6MA:bdlr3kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/33046e0ce784df9e05179c128629fcce659a9862/pkg/util/json/json.go#L687)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Doesn't this break the invariant that each key slice is sorted, and also leave us with duplicate keys? I could be misunderstanding, but it looks like the caller of this function creates a new key for every value in the returned slice. In that case, we'll be incorrectly emitting `nil`-suffixed keys.
</blockquote></details>

I just added a check for nil so that it skips them in the caller. Both arrays are sorted when they get to this function, but then it marks entries that overlap as nil in the second array which is still sorted, so that I don't need to allocate another array and everything still stays sorted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665)*

**Comment by jordanlewis** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/util/json/json.go, line 687 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3UjSrMgJg6FhXJtlam-r3-687:-L3V2HM6zsgjoTMVCLh5:b11mhn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/33046e0ce784df9e05179c128629fcce659a9862/pkg/util/json/json.go#L687)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I just added a check for nil so that it skips them in the caller. Both arrays are sorted when they get to this function, but then it marks entries that overlap as nil in the second array which is still sorted, so that I don't need to allocate another array and everything still stays sorted.
</blockquote></details>

Won't this algorithm eventually return stuff in non-sorted order, since we output things as we see it?

Consider the array [[10],[1]]

I think this will get output in non-sorted order which will break the assumptions of the caller.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665)*

**Comment by jordanlewis** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/util/json/json.go, line 687 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3UjSrMgJg6FhXJtlam-r3-687:-L3V3-J72YYq99YuG0cd:b-9bxa6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/33046e0ce784df9e05179c128629fcce659a9862/pkg/util/json/json.go#L687)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Won't this algorithm eventually return stuff in non-sorted order, since we output things as we see it?

Consider the array [[10],[1]]

I think this will get output in non-sorted order which will break the assumptions of the caller.
</blockquote></details>

To correctly implement the merge step, you need to avoid appending to `outKeys` until you've guaranteed that you've seen the smallest element of all of your input sorted sets.

I think this is the same as the "k-way merge" algorithm. We should crib it from somewhere. https://en.wikipedia.org/wiki/K-way_merge_algorithm

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665)*

**Comment by m-schneider** [comment]
Wouldn't that array already be sorted as [[1],[10]]?

**Comment by jordanlewis** [comment]
What do you mean by already sorted? I'm talking about when we call `EncodeInvertedIndexKeys` on the outermost array.

First, we call `EncodedInvertedIndexKeys` on `[10]`, which outputs a single key with `10` in it. `removeOverlappingKeys` does nothing to that output since the `prevChildren` slice is empty. We append the `10` key to `outKeys`.

Then, we call `EncodedInvertedIndexKeys` on `[1]`. It makes a single key with `1` in it. `removeOverlappingKeys` is invoked on this key and the `10` key, and does nothing since there are no duplicates. Then, we append the `1` key to `outKeys`.

We then return `outKeys`, which is composed of the `10` key and the `1` key in that order. That's not sorted and will therefore break our recursion invariant that return values are sorted.


There's a worse problem than this, actually. This algorithm won't properly deduplicate the array `[1,2,3,1]` because it doesn't consider all `k` of its inputs when deciding which to increment. That's why we need to use the k-way merge sort I mentioned :) or move back to your original idea to use a hashmap and file an issue to make this more efficient later.

Either way, please add some more test cases. It would be better if the tests noticed these issues.

File: pkg/util/json/json.go
Comment (vivekmenezes) [inline pkg/util/json/json.go]: It will be very valuable to understand what this function is meant to do. What are the constraints on the parameters and what is returned.

File: pkg/util/json/json.go
Comment (vivekmenezes) [inline pkg/util/json/json.go]: the iterators and the lengths above can be local to the for loop

File: pkg/util/json/json.go
Comment (vivekmenezes) [inline pkg/util/json/json.go]: do we have to special case when keys1[it1] == nil

File: pkg/sql/logictest/testdata/logic_test/inverted_index
Comment (vivekmenezes) [inline pkg/sql/logictest/testdata/logic_test/inverted_index]: It's quite surprising that code in util/json is being tested using logictests. Besides these logic tests are you able to write tests within the util/json package for this change? Thanks!

**Review by vivekmenezes — COMMENTED** [review]
I'm glad you've caught this.

**Comment by m-schneider** [comment]
If we do k way merge sort then we'll have to sort every single array to see if there are duplicates in it and then do a merge on the subpaths of duplicate keys. The runtime on something like that could be much worse than using a hashset even if you take the most basic example for an array of size n with no duplicates. You would have to sort it which would already give a runtime of nlogn where n is the total number of keys output while a hashset would give a runtime of n. If the JSON has nested arrays then it can get even worse. So I'm going back to the original hashset implementation.

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 292 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3U9AGirKpsX9m9dpTo-r4-292:-L3Yo6cyvg2FGVKALGRx:b82adxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/915e19c0a1d6465dac96becc06b60abdb13050ca/pkg/sql/logictest/testdata/logic_test/inverted_index#L292)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

It's quite surprising that code in util/json is being tested using logictests. Besides these logic tests are you able to write tests within the util/json package for this change? Thanks!
</blockquote></details>

I've moved this back out of json.go.

---

*[pkg/util/json/json.go, line 680 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-L3UjSrMgJg6FhXJtlam-r4-680:-L3YofPAYp2dL1FLWrIP:b-kavelf) ([raw file](https://github.com/cockroachdb/cockroach/blob/915e19c0a1d6465dac96becc06b60abdb13050ca/pkg/util/json/json.go#L680)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

It will be very valuable to understand what this function is meant to do. What are the constraints on the parameters and what is returned.
</blockquote></details>

I looked at this logic again and I think this sort of merge or merge sort is too complicated for questionable performance gains.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3Yfstip9D--kluP7dn:b-9bb3zj)*

**Comment by jordanlewis** [comment]
I agree that the runtime of the merge sort version might be worse than the hash set version, but it seems like a complicated question to precisely answer. Each path has a lot of duplicated data in it - the prefix of the path. Storing the paths into the hash set requires hashing each of the strings, which ends up being a lot of duplicate work. Finally, we'll have to allocate this new hash set on every insert/update/delete of a json column, which will be O(# of paths) in space.

Additionally, you will have to do this unconditionally across the entire object, instead of just for paths that go through arrays.

The k-way merge version is more efficient in space. It does not have to allocate any extra memory. Its time complexity in the worst case is worse, but the fact that it can skip doing work for paths that don't go through an array with size greater than 1 seems like a huge win to me.


Anyway, we can argue all day about the estimated complexities but we won't really know one way or the other until we write some benchmarks. I'm happy to see this get merged with the hash set implementation, but I think we should file an issue to write benchmarks for key encoding and to investigate alternative approaches.

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3YpS2DkXe7srMngTEo:b-r0x4dv)*

File: pkg/sql/sqlbase/table.go
Comment (vivekmenezes) [inline pkg/sql/sqlbase/table.go]:
```
for i, byteValue := range keySet {
 +			outKeys[i] = byteValue
}
```

File: pkg/sql/sqlbase/table.go
Comment (vivekmenezes) [inline pkg/sql/sqlbase/table.go]: Add a comment here that you're getting rid of duplicate keys

Comment (vivekmenezes) [reply pkg/sql/sqlbase/table.go]: Is it possible to convert EncodeInvertedIndexKeys() to not return duplicates in the first place?

**Review by vivekmenezes — COMMENTED** [review]
Thanks for clarifying what this change is doing

File: pkg/sql/sqlbase/table.go
Comment (jordanlewis) [reply pkg/sql/sqlbase/table.go]: Converting it to not return duplicates in the first place is what the discussion about k-way merge is about.

**Comment by jordanlewis** [comment]
Good point! There still might be some tradeoff here, because it's costly to assemble a giant batch full of duplicates and send it potentially over the wire, but I'm happier to see that particular can kicked down the road.

---

Review status: 0 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21665#-:-L3dcHHr-tns7__WiY_L:b-z6i3w0)*

**Review by jordanlewis — APPROVED** [review]
LGTM mod commented out code!

**Pull request merged**
